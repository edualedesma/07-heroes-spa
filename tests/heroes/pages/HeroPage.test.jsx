import { fireEvent, render, screen } from "@testing-library/react";
import { MemoryRouter, Route, Routes } from "react-router-dom";
import { HeroPage } from "../../../src/heroes/pages/HeroPage";

const mockedUseNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockedUseNavigate,
}));

describe('Testing on <HeroPage />', () => {

    beforeEach( () => jest.clearAllMocks() );

    test('should show the initial state of the component', () => {

        const { container } = render(
            <MemoryRouter initialEntries={['/hero/dc-superman']}>
                <Routes>
                    <Route path='hero/:id' element={ <HeroPage /> } />
                </Routes>
            </MemoryRouter>
        );

        expect( container ).toMatchSnapshot();
        // screen.debug();

    });

    test('should show superman hero card', () => {

        render(
            <MemoryRouter initialEntries={['/hero/dc-superman']}>
                <Routes>
                    <Route path='hero/:id' element={ <HeroPage /> } />
                </Routes>
            </MemoryRouter>
        );

        const img = screen.getByRole('img');
        expect( img.alt ).toBe('Superman');
        expect( screen.getByText('Superman') ).toBeTruthy();
        // screen.debug();

    });

    test('should redirect to "/marvel" page if hero does not exist', () => {

        render(
            <MemoryRouter initialEntries={['/hero/dc-supermannn']}>
                <Routes>
                    <Route path='hero/:id' element={ <HeroPage /> } />
                    <Route path='/marvel' element={ <h1>Marvel Page</h1> } />
                </Routes>
            </MemoryRouter>
        );

        // expect( mockedUseNavigate ).toHaveBeenCalled();
        expect( screen.getByText('Marvel Page') ).toBeTruthy();
        // screen.debug();
    });

    test('should redirect to back page if click on back button', () => {

        render(
            <MemoryRouter initialEntries={['/hero/dc-superman']}>
                <Routes>
                    <Route path='hero/:id' element={ <HeroPage /> } />
                </Routes>
            </MemoryRouter>
        );

        const btn = screen.getByRole('button');
        fireEvent.click( btn );

        expect( mockedUseNavigate ).toHaveBeenCalled();
        // screen.debug();
    });

});