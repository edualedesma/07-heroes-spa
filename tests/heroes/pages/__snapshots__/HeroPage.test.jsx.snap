// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing on <HeroPage /> should show the initial state of the component 1`] = `
<div>
  <div
    class="row mt-5"
  >
    <div
      class="col-4"
    >
      <img
        alt="Superman"
        class="img-thumbnail animate__animated animate__fadeInLeft"
        src="/heroes/dc-superman.jpg"
      />
    </div>
    <div
      class="col-8"
    >
      <h3>
        Superman
      </h3>
      <ul
        class="list-group list-group-flush"
      >
        <li
          class="list-group-item"
        >
          <b>
            Alter ego:
          </b>
           
          Kal-El
        </li>
        <li
          class="list-group-item"
        >
          <b>
            Publisher:
          </b>
           
          DC Comics
        </li>
        <li
          class="list-group-item"
        >
          <b>
            First appearance:
          </b>
           
          Action Comics #1
        </li>
      </ul>
      <h5
        class="mt-3"
      >
        Characters
      </h5>
      <p>
        Kal-El
      </p>
      <button
        class="btn btn-outline-primary"
      >
        Atr√°s
      </button>
    </div>
  </div>
</div>
`;
